pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -415,6 +423,106 @@ func (r *ReconcileClusterPool) Reconcile(ctx context.Context, request reconcile.
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileClusterPool) updateInventory(clp *hivev1.ClusterPool, cds []*hivev1.ClusterDeployment, valid bool, state string, logger log.FieldLogger) {
+	if clp.Spec.Inventory != nil {
+		var active_cdc []string
+		for _, cd := range cds {
+			cdcRef := cd.Spec.ClusterPoolRef.ClusterDeploymentCustomizationRef
+			active_cdc = append(active_cdc, cdcRef.Name)
+			cdc := &hivev1.ClusterDeploymentCustomization{}
+			if err := r.Client.Get(context.Background(), client.ObjectKey{Namespace: clp.Namespace, Name: cdcRef.Name}, cdc); err != nil {
+				if apierrors.IsNotFound(err) {
+					r.setInventoryValidCondition(clp, true, cdc.Name, "missing", logger)
```

Comment:
> However, there's another case to consider: If someone deleted the CDC from the Inventory, any status we report for it in the pool condition will look odd. Perhaps we need to distinguish between "CDC does not exist in the database" and "CDC is absent from the inventory, despite being referenced by a CD belonging to this ClusterPool". I'm not sure what that would look like. Suggestions welcome

I would prefer that the relevant CD will be marked broken and will be deleted automatically by CP.

Reply:
Per the [design](https://github.com/openshift/hive/blob/master/docs/enhancements/clusterpool-inventory.md#removing-an-entry-from-the-inventory), deleting depends on whether the CD is claimed or not. But it's not really "broken" in either case.

#### Tangent: How to do that deletion
When we discover an **unclaimed** CD whose inventory entry is gone, we need to delete it. However, we can't just issue a `Delete()` because we have `MaxConcurrent` to worry about ([doc](https://github.com/openshift/hive/blob/bfb69aec0a7206918012bb21ed25cae098401b59/apis/hive/v1/clusterpool_types.go#L36-L40)). TL;DR: the clusterpool controller must handle all deletions of CDs belonging to a pool, and must do so while keeping the total number of active provisions+deprovisions at or below `MaxConcurrent`.

To handle this for deleting a claimed cluster when its ClusterClaim is deleted, we have the concept of "marked for deletion". The clusterclaim controller [adds an annotation to the CD](https://github.com/openshift/hive/blob/bfb69aec0a7206918012bb21ed25cae098401b59/pkg/controller/clusterclaim/clusterclaim_controller.go#L413-L424), and then the clusterpool controller [does the actual `Delete()`](https://github.com/openshift/hive/blob/bfb69aec0a7206918012bb21ed25cae098401b59/pkg/controller/clusterpool/clusterpool_controller.go#L342-L353) within the bounds of `MaxConcurrent` (tracked via the local variable `availableCurrent`).

For deleting a CD whose inventory entry has gone missing, I suggest we piggyback on that. Rather trying to add and track a new annotation, let's just make the existing one more generic and use it in both places. I've opened https://github.com/openshift/hive/pull/1723 to do the refactor work. Once that merges, you should be able to do something just like [this](https://github.com/openshift/hive/blob/8490d7ca6cb4c5bf14bc8f28b1287b8ceaf19c55/pkg/controller/clusterclaim/clusterclaim_controller.go#L412-L420) in the code path where you discover a missing inventory entry for an unclaimed CD.

Note that the design says we should *not* delete a **claimed** CD whose inventory entry has been removed.

#### Back to our regularly scheduled comment

...all of which still leaves the question of whether/how we should mark the status of such a CD, whether claimed or unclaimed, and/or its ClusterPool.

I think I'm okay leaving the CD status alone. We don't really want to get in the business fo trying to update CD status every time something about its inventory entry changes. It would be a fair bit of extra code to write, test, and maintain, and I don't see a strong use case for it.

For the pool, let's split `"missing"` into `"cdc missing"` and `"inventory entry missing"`. (Briefly considered including the name of the CD in the latter case; but I guess that's not really necessary. For any of these you can look up the CD based on the CDC name.)